src/object_store/src/object/opendal_engine/opendal_object_store.rs

```diff
@@ -195,34 +193,37 @@ impl ObjectStore for OpendalObjectStore {
 
 /// Store multiple parts in a map, and concatenate them on finish.
 pub struct OpenDalStreamingUploader {
-    op: Operator,
-    path: String,
-    buffer: BytesMut,
+    writer: Writer,
 }
 impl OpenDalStreamingUploader {
-    pub fn new(op: Operator, path: String) -> Self {
-        Self {
-            op,
-            path,
-            buffer: BytesMut::new(),
-        }
+    pub async fn new(op: Operator, path: String) -> ObjectResult<Self> {
+        let writer = op.writer(&path).await?;
+        Ok(Self { writer })
     }
 }
+
+unsafe impl Sync for OpenDalStreamingUploader {}
+
+const OPENDAL_BUFFER_SIZE: u64 = 8 * 1024 * 1024;
 #[async_trait::async_trait]
 impl StreamingUploader for OpenDalStreamingUploader {
     async fn write_bytes(&mut self, data: Bytes) -> ObjectResult<()> {
-        self.buffer.put(data);
+        self.writer.write(data).await?;
         Ok(())
     }
 
     async fn finish(mut self: Box<Self>) -> ObjectResult<()> {
-        self.op.write(&self.path, self.buffer).await?;
+        match self.writer.close().await {
+            Ok(_) => (),
+            Err(_) => self.writer.abort().await?,
+        };
 
         Ok(())
     }
 
     fn get_memory_usage(&self) -> u64 {
-        self.buffer.capacity() as u64
+        // todo: refactor this after https://github.com/apache/incubator-opendal/issues/2087 is implemented.
+        OPENDAL_BUFFER_SIZE
```

Comment:
Will use configurable buffer size after OpenDAL bump new version.

Reply:
Oh forget this, since we have not bench different part size, I guess  it's ok to use the fixed buffer size(8MB) as default. 
for gcs, [it's recommended that you use at least 8 MiB for the chunk size](https://cloud.google.com/storage/docs/performing-resumable-uploads);
for oss, [the part size is 100 KB to 5 GB](https://www.alibabacloud.com/help/en/object-storage-service/latest/multipart-upload-6).

Reply:
In the future, if we want to config buffer size, we need to set it before building an engine, and maybe this is a storage/object_store config. For example for gcs:
```
        builder.write_fixed_size(8*1024*1024);
        let op: Operator = Operator::new(builder)?.finish();
```